include/solvers/mpm_base.tcc

```diff
@@ -338,7 +338,9 @@ void mpm::MPMBase<Tdim>::initialise_particle_types() {
     // Gather particle types
     auto particle_type =
         json_particle["generator"]["particle_type"].template get<std::string>();
-    particle_types_.insert(particle_type);
+    // Insert only unique particle types
+    if (particle_types_.find(particle_type) == particle_types_.end())
+      particle_types_.insert(particle_type);
```

Comment:
I thought std::set only stores unique entries, i.e., it does not store duplicates. If so, then this check is unnecessary.

Reply:
Maybe you're right, documentation seems to indicate that. I thought when I was testing it I saw it go through P2D twice though. Will confirm and remove if unnecessary.